File: src/services/weather.service.js
/**
 * Weather Service — FarmEasy Krishi Raksha
 * Uses Open-Meteo (https://open-meteo.com) — completely FREE, no API key needed.
 * Uses Nominatim (OpenStreetMap) for pincode → coordinates — also FREE.
 *
 * Replaces OpenWeatherMap so OPENWEATHER_API_KEY is no longer required.
 */
import axios from 'axios';

const OPEN_METEO  = 'https://api.open-meteo.com/v1';
const NOMINATIM   = 'https://nominatim.openstreetmap.org';

// ─── WMO Weather Interpretation Code → human description ─────────────────────
function wmoDesc(code) {
  if (code === 0)              return 'clear sky';
  if (code <= 3)               return 'partly cloudy';
  if (code <= 48)              return 'foggy';
  if (code <= 55)              return 'drizzle';
  if (code <= 65)              return 'rain';
  if (code <= 77)              return 'snow';
  if (code <= 82)              return 'rain showers';
  if (code <= 86)              return 'snow showers';
  if (code >= 95)              return 'thunderstorm';
  return 'overcast';
}

// ─── Pincode → Coordinates ────────────────────────────────────────────────────
// Primary: Nominatim (OpenStreetMap geocoding — free, no key)
// Fallback: hardcoded Maharashtra district centres
async function pincodeToCoords(pincode) {
  try {
    const { data } = await axios.get(`${NOMINATIM}/search`, {
      params: { postalcode: pincode, country: 'IN', format: 'json', limit: 1 },
      headers: { 'User-Agent': 'FarmEasy/1.0' },
      timeout: 6000,
    });
    if (data && data.length > 0) {
      return {
        lat: parseFloat(data[0].lat),
        lon: parseFloat(data[0].lon),
        name: data[0].display_name.split(',')[0].trim(),
      };
    }
  } catch { /* fall through to hardcoded */ }

  // Hardcoded Maharashtra fallback keyed by first 3 digits of pincode
  const fallback = {
    '413': { lat: 17.68, lon: 75.28, name: 'Solapur' },
    '411': { lat: 18.52, lon: 73.86, name: 'Pune' },
    '400': { lat: 19.07, lon: 72.87, name: 'Mumbai' },
    '422': { lat: 19.99, lon: 73.79, name: 'Nashik' },
    '431': { lat: 19.87, lon: 75.34, name: 'Aurangabad' },
    '440': { lat: 21.14, lon: 79.09, name: 'Nagpur' },
    '416': { lat: 16.70, lon: 74.24, name: 'Kolhapur' },
    '415': { lat: 17.00, lon: 74.30, name: 'Sangli' },
    '425': { lat: 21.00, lon: 74.74, name: 'Dhule' },
    '444': { lat: 20.59, lon: 78.12, name: 'Amravati' },
    '423': { lat: 20.55, lon: 74.38, name: 'Malegaon' },
    '414': { lat: 19.08, lon: 75.01, name: 'Ahmednagar' },
    '432': { lat: 20.00, lon: 76.14, name: 'Jalna' },
    '431': { lat: 19.87, lon: 75.34, name: 'Chhatrapati Sambhajinagar' },
  };
  const prefix = String(pincode).slice(0, 3);
  return fallback[prefix] || { lat: 18.52, lon: 73.86, name: 'Maharashtra' };
}

// ─── Magnus formula for dew point ─────────────────────────────────────────────
function calcDewPoint(tempC, humidity) {
  const a = 17.27, b = 237.7;
  const alpha = ((a * tempC) / (b + tempC)) + Math.log(humidity / 100);
  return Math.round((b * alpha) / (a - alpha) * 10) / 10;
}

// ─── Disease risk assessment ───────────────────────────────────────────────────
function assessWeatherRisk(current, forecast) {
  const risks = [];
  const next3 = forecast.slice(0, 3);

  const avgHumidity = next3.reduce((a, b) => a + b.humidity, 0) / next3.length;
  const avgTemp     = next3.reduce((a, b) => a + b.tempAvg,  0) / next3.length;
  const totalRain   = next3.reduce((a, b) => a + b.rainfall, 0);
  const highCloudDays = next3.filter((d) => d.cloudCover > 70).length;

  if (avgHumidity > 80 && avgTemp >= 24 && avgTemp <= 32) {
    risks.push('HIGH fungal disease risk — humidity >80% + temp 24–32°C ideal for Blast/Blight sporulation');
  } else if (avgHumidity > 70) {
    risks.push('MODERATE fungal disease risk — elevated humidity favours mildew and rust');
  }
  if (totalRain > 20) {
    risks.push('Heavy rainfall — risk of waterlogging, root rot, and splash-dispersed diseases');
  } else if (totalRain > 5) {
    risks.push('Moderate rainfall — conditions favour bacterial blight spread');
  }
  if (highCloudDays >= 2) risks.push('Extended cloud cover — reduced UV, increased leaf wetness duration');
  if (current.windSpeed > 30) risks.push('High wind speed — spore dispersal risk elevated');

  const riskLevel =
    risks.some((r) => r.startsWith('HIGH')) ? 'HIGH' :
    risks.length > 1 ? 'MODERATE' : 'LOW';

  return {
    riskLevel,
    riskFactors: risks,
    avgHumidity: Math.round(avgHumidity),
    avgTemp: Math.round(avgTemp),
    totalRainfall3Days: totalRain,
    leafWetnessPeriod: avgHumidity > 75 ? 'PROLONGED (>8 hrs/day)' : 'NORMAL (<6 hrs/day)',
  };
}

// ─── Main export ──────────────────────────────────────────────────────────────
export async function getWeatherData(pincode) {
  const coords = await pincodeToCoords(pincode);

  const { data } = await axios.get(`${OPEN_METEO}/forecast`, {
    params: {
      latitude:  coords.lat,
      longitude: coords.lon,
      current: [
        'temperature_2m', 'relative_humidity_2m', 'apparent_temperature',
        'wind_speed_10m', 'weather_code', 'cloud_cover',
        'surface_pressure', 'dew_point_2m',
      ].join(','),
      daily: [
        'temperature_2m_max', 'temperature_2m_min',
        'precipitation_sum', 'weather_code',
        'relative_humidity_2m_max', 'cloud_cover_mean',
      ].join(','),
      forecast_days:   5,
      timezone:        'Asia/Kolkata',
      wind_speed_unit: 'kmh',
    },
    timeout: 10000,
  });

  const c = data.current;
  const current = {
    temp:        Math.round(c.temperature_2m),
    tempMin:     Math.round(c.temperature_2m),
    tempMax:     Math.round(c.temperature_2m),
    feelsLike:   Math.round(c.apparent_temperature),
    humidity:    c.relative_humidity_2m,
    pressure:    Math.round(c.surface_pressure),
    windSpeed:   Math.round(c.wind_speed_10m),
    windDeg:     0,
    cloudCover:  c.cloud_cover,
    visibility:  10,
    weatherMain: wmoDesc(c.weather_code),
    weatherDesc: wmoDesc(c.weather_code),
    uvIndex:     null,
    dewPoint:    Math.round((c.dew_point_2m ?? calcDewPoint(c.temperature_2m, c.relative_humidity_2m)) * 10) / 10,
    timestamp:   new Date().toISOString(),
  };

  const d = data.daily;
  const forecast = d.time.map((date, i) => ({
    date,
    tempMin:    Math.round(d.temperature_2m_min[i]),
    tempMax:    Math.round(d.temperature_2m_max[i]),
    tempAvg:    Math.round((d.temperature_2m_min[i] + d.temperature_2m_max[i]) / 2),
    humidity:   d.relative_humidity_2m_max?.[i] ?? 60,
    rainfall:   Math.round((d.precipitation_sum[i] || 0) * 10) / 10,
    cloudCover: Math.round(d.cloud_cover_mean?.[i] || 0),
    description: wmoDesc(d.weather_code[i]),
  }));

  const weatherRisk = assessWeatherRisk(current, forecast);

  return {
    location: { pincode, name: coords.name, lat: coords.lat, lon: coords.lon },
    current,
    forecast,
    weatherRisk,
    fetchedAt: new Date().toISOString(),
  };
}

File: AI_CROP_DISESE_DETECTION/weather_service.py
"""
Weather Service — Open-Meteo (free, no API key required)
Fetches current conditions + 7-day forecast + agriculture variables.
"""
from __future__ import annotations
import logging
import httpx
from typing import Optional

logger = logging.getLogger(__name__)


_CURRENT_VARS = (
    "temperature_2m,relative_humidity_2m,apparent_temperature,"
    "precipitation,wind_speed_10m,cloud_cover,weather_code,"
    "dew_point_2m,vapour_pressure_deficit,et0_fao_evapotranspiration"
)
_DAILY_VARS = (
    "temperature_2m_max,temperature_2m_min,precipitation_sum,"
    "relative_humidity_2m_max,relative_humidity_2m_min,wind_speed_10m_max,"
    "precipitation_probability_max,weather_code"
)
_HOURLY_SOIL = (
    "soil_temperature_0cm,soil_temperature_6cm,"
    "soil_moisture_0_to_1cm,soil_moisture_1_to_3cm,soil_moisture_3_to_9cm"
)

WEATHER_CODES = {
    0: "Clear sky", 1: "Mainly clear", 2: "Partly cloudy", 3: "Overcast",
    45: "Fog", 48: "Depositing rime fog",
    51: "Light drizzle", 53: "Moderate drizzle", 55: "Dense drizzle",
    61: "Slight rain", 63: "Moderate rain", 65: "Heavy rain",
    71: "Slight snow", 73: "Moderate snow", 75: "Heavy snow",
    80: "Slight showers", 81: "Moderate showers", 82: "Violent showers",
    95: "Thunderstorm", 96: "Thunderstorm with hail", 99: "Thunderstorm with heavy hail",
}


def _risk_from_weather(current: dict, daily: list[dict]) -> dict:
    """Derive a basic disease-risk flag from raw weather data."""
    temp   = current.get("temperature_2m", 20)
    rh     = current.get("relative_humidity_2m", 60)
    vpd    = current.get("vapour_pressure_deficit", 1.0)
    precip = current.get("precipitation", 0)

    risk_factors = []
    if rh > 80 and 15 <= temp <= 30:
        risk_factors.append(f"High humidity ({rh}%) + warm temp favours fungal disease")
    if rh > 90:
        risk_factors.append("Critical humidity (>90%) — leaf wetness risk")
    if precip > 0:
        risk_factors.append("Recent precipitation — splash dispersal risk for bacteria/fungi")
    if vpd < 0.4:
        risk_factors.append("Low VPD — high leaf wetness duration")
    if temp > 35 and rh < 40:
        risk_factors.append("High temp + dry conditions — spider mite risk")

    # Check 3-day forecast for upcoming rain
    rain_days = sum(1 for d in daily[:3] if d.get("precipitation_sum", 0) > 2)
    if rain_days >= 2:
        risk_factors.append(f"Rain forecast {rain_days}/3 days — sustained disease-favourable window")

    if len(risk_factors) >= 3 or rh > 90:
        risk_level = "HIGH"
    elif len(risk_factors) >= 2 or rh > 80:
        risk_level = "MODERATE"
    elif risk_factors:
        risk_level = "LOW"
    else:
        risk_level = "LOW"

    return {"risk_level": risk_level, "risk_factors": risk_factors}


async def fetch_weather(latitude: float, longitude: float) -> Optional[dict]:
    """
    Returns a structured weather dict or None on failure.
    """
    url = "https://api.open-meteo.com/v1/forecast"
    params = {
        "latitude": latitude,
        "longitude": longitude,
        "current": _CURRENT_VARS,
        "daily": _DAILY_VARS,
        "hourly": _HOURLY_SOIL,
        "timezone": "Asia/Kolkata",
        "forecast_days": 7,
    }

    try:
        async with httpx.AsyncClient(timeout=10) as client:
            resp = await client.get(url, params=params)
            resp.raise_for_status()
            raw = resp.json()

        current_raw = raw.get("current", {})
        daily_raw   = raw.get("daily", {})
        hourly_raw  = raw.get("hourly", {})

        # Build daily forecast list
        dates    = daily_raw.get("time", [])
        daily_list = []
        for i, date in enumerate(dates):
            daily_list.append({
                "date": date,
                "temp_max":  daily_raw.get("temperature_2m_max",  [None]*10)[i],
                "temp_min":  daily_raw.get("temperature_2m_min",  [None]*10)[i],
                "rainfall":  daily_raw.get("precipitation_sum",   [0]*10)[i],
                "humidity_max": daily_raw.get("relative_humidity_2m_max", [None]*10)[i],
                "humidity_min": daily_raw.get("relative_humidity_2m_min", [None]*10)[i],
                "wind_max":  daily_raw.get("wind_speed_10m_max",  [None]*10)[i],
                "rain_prob": daily_raw.get("precipitation_probability_max", [0]*10)[i],
                "weather_code": daily_raw.get("weather_code", [0]*10)[i],
                "weather_desc": WEATHER_CODES.get(daily_raw.get("weather_code", [0]*10)[i], "Unknown"),
            })

        current = {
            "temperature":          current_raw.get("temperature_2m"),
            "humidity":             current_raw.get("relative_humidity_2m"),
            "apparent_temperature": current_raw.get("apparent_temperature"),
            "precipitation":        current_raw.get("precipitation", 0),
            "wind_speed":           current_raw.get("wind_speed_10m"),
            "cloud_cover":          current_raw.get("cloud_cover"),
            "weather_code":         current_raw.get("weather_code", 0),
            "weather_desc":         WEATHER_CODES.get(current_raw.get("weather_code", 0), "Unknown"),
            "dew_point":            current_raw.get("dew_point_2m"),
            "vpd":                  current_raw.get("vapour_pressure_deficit"),
            "evapotranspiration":   current_raw.get("et0_fao_evapotranspiration"),
        }

        # Latest soil readings (index 0 = current hour)
        soil = {
            "temperature_surface": (hourly_raw.get("soil_temperature_0cm") or [None])[0],
            "temperature_6cm":     (hourly_raw.get("soil_temperature_6cm") or [None])[0],
            "moisture_0_1cm":      (hourly_raw.get("soil_moisture_0_to_1cm") or [None])[0],
            "moisture_1_3cm":      (hourly_raw.get("soil_moisture_1_to_3cm") or [None])[0],
            "moisture_3_9cm":      (hourly_raw.get("soil_moisture_3_to_9cm") or [None])[0],
        }

        weather_risk = _risk_from_weather(current_raw, daily_list)

        return {
            "location": {"latitude": latitude, "longitude": longitude},
            "current": current,
            "daily_forecast": daily_list,
            "soil": soil,
            "weather_risk": weather_risk,
        }

    except Exception as exc:
        logger.warning(f"[WeatherService] Failed to fetch weather: {exc}")
        return None
